### DISCUSSION TITLE
structuralSharing should be false by default

### DISCUSSION THREAD

**Discussion by fermmm** [TanStack/query#1709] [Ideas]
The issues received by the users when structuralSharing did not exist are this kind of issues "I want this feature", example:
https://github.com/tannerlinsley/react-query/issues/430

The issues received by the users when structuralSharing was implemented and set to true by default are this kind of issues: "I have a bug and I don't know what is happening", examples:
https://github.com/tannerlinsley/react-query/issues/1633
https://github.com/tannerlinsley/react-query/issues/1657#issuecomment-769456647

"I want this feature" vs "I have a bug"... I think it's easy to decide, having a bug consumes a lot of time. Features that are enabled by default should be the ones that does not generate hard to debug issues or "gotchas". Also the warning for this is well hidden on the documentation, which is not friendly.

**Comment by tannerlinsley** [comment]
We're always open to improving the documentation, but for now the plan is to stick with the default structuralSharing. The side-effects of turning this off could be much worse than most anticipate. The expectation for data structures to remain constant from render to render is a very very common one in the React ecosystem, which was the first motivator for this feature, among many others. Removing this default would effectively start mal-triggering every single `useEffect` or `===` change detection mechanism in every single app that uses React Query.

Converting this to a discussion so we can continue to keep tabs and discuss it.

**Reply by CoryDanielson** [reply]
Has there been any thought to exposing `replaceEqualDeep` as an option so that users could write a custom version specific to their use case/data? Also what about allowing it to be asynchronous so that the operation could be chunked so that it doesn't have as much of an impact on very large datasets?

**Comment by ngokevin** [comment]
For retro, I also had an issue with structuralSharing that ate up a day of debugging: https://github.com/TanStack/query/issues/6812
